slowquant/qiskit_interface/operators_circuits.py

```diff
@@ -1,165 +1,104 @@
-        «                └───┘     └───┘     ┌─┴─┐└─────────┘└────────┘  │  »
-        «q_2: ───────────────────────────────┤ X ├───────────────────────┼──»
-        «                                    └───┘                     ┌─┴─┐»
-        «q_3: ─────────────────────────────────────────────────────────┤ X ├»
-        «                                                              └───┘»
-        «     ┌─────────────┐     ┌─────────┐      ┌────────┐
-        «q_0: ┤ Ry(-0.25*θ) ├──■──┤ Ry(θ/4) ├──■───┤ P(π/2) ├──────────────■───────
-        «     └─────────────┘  │  └─────────┘┌─┴─┐┌┴────────┤┌──────────┐  │
-        «q_1: ─────────────────┼─────────────┤ X ├┤ P(-π/2) ├┤ Ry(-π/2) ├──┼────■──
-        «                    ┌─┴─┐   ┌───┐   ├───┤└─────────┘└──────────┘┌─┴─┐  │
-        «q_2: ───────────────┤ X ├───┤ H ├───┤ X ├───────────────────────┤ X ├──┼──
-        «          ┌───┐     ├───┤   └───┘   └───┘                       └───┘┌─┴─┐
-        «q_3: ─────┤ H ├─────┤ X ├────────────────────────────────────────────┤ X ├
-        «          └───┘     └───┘                                            └───┘
-
-    #. 10.48550/arXiv.2312.09761
-    #. 10.1103/PhysRevA.102.062612, Fig. 7
-    #. 10.1038/s42005-021-00730-0, Fig. 2
+def sa_single_excitation(
+    i: int,
+    a: int,
+    num_orbs: int,
+    qc: QuantumCircuit,
+    theta: Parameter | ParameterExpression,
+    mapper: FermionicMapper,
+) -> QuantumCircuit:
+    """Get spin-adapted single singlet excitation circuit.
 
     Args:
-        p: Occupied spatial index.
+        i: Strongly occupied spatial orbital index.
+        a: Weakly occupied spatial orbital index.
         num_orbs: Number of spatial orbitals.
         qc: Quantum circuit.
         theta: Circuit parameter.
+        mapper: Fermionic to qubit mapper.
 
     Returns:
-        Modified quantum circuit.
+        Spin-adpated singlet single excitation circuit.
     """
-    l = p
-    j = p + 1
-    k = p + num_orbs
-    i = p + num_orbs + 1
-    qc.cx(l, k)
-    qc.cx(j, i)
-    qc.cx(l, j)
-    qc.x(k)
-    qc.x(i)
-    qc.ry(-theta / 4, l)
-    qc.h(k)
-    qc.cx(l, k)
-    qc.ry(theta / 4, l)
-    qc.h(i)
-    qc.cx(l, i)
-    qc.ry(-theta / 4, l)
-    qc.cx(l, k)
-    qc.ry(theta / 4, l)
-    qc.h(j)
-    qc.cx(l, j)
-    qc.ry(-theta / 4, l)
-    qc.cx(l, k)
-    qc.ry(theta / 4, l)
-    qc.cx(l, i)
-    qc.ry(-theta / 4, l)
-    qc.h(i)
-    qc.cx(l, k)
-    qc.ry(theta / 4, l)
-    qc.h(k)
-    qc.h(j)
-    qc.ry(np.pi / 2, j)
-    qc.p(np.pi / 2, j)
-    qc.cx(l, j)
-    qc.p(np.pi / 2, l)
-    qc.p(-np.pi / 2, j)
-    qc.ry(-np.pi / 2, j)
-    qc.x(k)
-    qc.x(i)
-    qc.cx(j, i)
-    qc.cx(l, k)
+    if isinstance(mapper, JordanWignerMapper):
+        qc = sa_single_excitation_efficient(a, i, num_orbs, qc, theta)
+    else:
+        qc = sa_single_excitation_trotter(i, a, num_orbs, qc, theta, mapper)
     return qc
 
 
-def single_excitation(
+def single_excitation_efficient(
```

Comment:
Maybe all the circuit building functions that are called via the general function that checks for JW or not could be called _<name>. So as private?